### ISSUE TITLE
ExternalTaskSensor requires an exact execution_delta

### ISSUE THREAD

**Issue by jtschoonhoven** [apache/airflow#1060]
When poking an external queue, ExternalTaskSensor only matches a task_instance with an `execution_date` that _exactly_ matches the current `execution_date` or `execution_date - execution_delta`. If a task in DAG A waits on a task in DAG B and DAG A gets rescheduled, this dependency will break.

```
        count = session.query(TI).filter(
            TI.dag_id == self.external_dag_id,
            TI.task_id == self.external_task_id,
            TI.state.in_(self.allowed_states),
            TI.execution_date == dttm,
        ).count()
```

https://github.com/airbnb/airflow/blob/master/airflow/operators/sensors.py#L196

_Proposal_
Change the above to `TI.execution_date >= dttm` so that we can, for example, poke for an external task to have completed since UTC midnight without having to know the exact execution date.

Happy to PR this.

**Comment by mistercrunch** [comment]
You're assuming that the ETL is non incremental which might be right for your pipeline but certainly isn't right for all use case. If I'm doing incremental loads and currently loading the 15th of the month, having some external job succeed for the 16th doesn't help me with my data for the 15th. Yes the job for the 16th can move forward, but the one for the 15th should still wait for its dependency.

**Comment by jtschoonhoven** [comment]
That makes sense. Would still be nice to have the option though. What about adding another kwarg to ExternalTaskSensor that makes it an option? Or would that make the API too messy?

**Comment by garaujodealmeida** [comment]
It is not possible one dag depends on other dag without time setting exactly deltatime?

**Comment by omnilinguist** [comment]
@jlowin @mistercrunch @criccomini @r39132

[TODO migrate to JIRA when lockdown lifted]

I'm unable to post on a corresponding JIRA ticket (https://issues.apache.org/jira/browse/AIRFLOW-100, https://issues.apache.org/jira/browse/AIRFLOW-93) due to lockdown, but I have successfully monkey-patched `ExternalTaskSensor` with a class that looks like this that should be reasonably flexible, what do you think? I tested it locally and it works as expected. Open to feedback on any recommended modifications.

Copied from gitter:

> is anyone else running into issues with the ExternalTaskSensor requiring an exact execution_date ( apache/incubator-airflow#1060 )?
> this kind of functionality works well only when the jobs are being run at well-known times
> but suppose now that one has some tasks in a DAG dependent on those in another DAG (apache/incubator-airflow#6), and maybe he wants to also enable such dependencies to work with triggered runs, whose execution_dates will not fit so nicely into predetermined time intervals
> would it not make sense to perhaps have some sort of +/- tolerance interval on each side of the target date difference to support these kind of "fuzzy queries"?

(Edit: Realised after the fact looking at the JIRA that an even more general version of this could be passing in an arity-1 function that will get passed the `context['execution_date']` within the scope and is expected to return a tuple `(start_date, end_date)` that can then be used in the SQLAlchemy query.)

``` python
class TimeRangeExternalTaskSensor(ExternalTaskSensor):
    """
    Monkey patch ExternalTaskSensor to enable fuzzy date searching; this is a hack
    :param execution_date_tolerance_before: how long before the target execution time to search for a task, in seconds
    :type execution_date_tolerance_before: int
    :param execution_date_tolerance_after: how long after the target execution time to search for a task, in seconds
    :type execution_date_tolerance_after: int
    """

    def __init__(
            self,
            execution_date_tolerance_before=0,
            execution_date_tolerance_after=0,
            *args, **kwargs):
        super(TimeRangeExternalTaskSensor, self).__init__(*args, **kwargs)
        self.execution_date_tolerance_before = execution_date_tolerance_before
        self.execution_date_tolerance_after = execution_date_tolerance_after

    def poke(self, context):
        if self.execution_delta:
            dttm = context['execution_date'] - self.execution_delta
        else:
            dttm = context['execution_date']

        logging.info(
            'Poking for '
            '{self.external_dag_id}.'
            '{self.external_task_id} on '
            '{dttm} ... '.format(**locals()))
        TI = TaskInstance

        execution_date_lower_bound = dttm - datetime.timedelta(seconds=self.execution_date_tolerance_before)
        execution_date_upper_bound = dttm + datetime.timedelta(seconds=self.execution_date_tolerance_after)
        logging.info('Searching for matching task(s) with execution_date between [{}, {}]...'.format(execution_date_lower_bound, execution_date_upper_bound))

        session = settings.Session()
        count = session.query(TI).filter(
            TI.dag_id == self.external_dag_id,
            TI.task_id == self.external_task_id,
            TI.state.in_(self.allowed_states),
            TI.execution_date >= execution_date_lower_bound,
            TI.execution_date <= execution_date_upper_bound
        ).count()
        session.commit()
        session.close()
        return count
```
